File: 2841.maximum-sum-of-almost-unique-subarray.py
#
# @lc app=leetcode id=2841 lang=python3
#
# [2841] Maximum Sum of Almost Unique Subarray
#

# @lc code=start
from collections import Counter
class Solution:
    def maxSum(self, nums: list[int], m: int, k: int) -> int:
        sum = 0
        sum_list = []
        counter_dict = Counter()
        for i,c in enumerate(nums):
            sum += c 
            counter_dict[c] += 1
            if i <  k-1:
                continue
            if len(counter_dict) < m:
                sum -= nums[i - k +1]
                counter_dict[nums[i - k +1]] -= 1
                if counter_dict[nums[i - k +1]] == 0:
                    del counter_dict[nums[i - k +1]]
                continue
            sum_list.append(sum)
            #Exit the window
            sum -= nums[i - k +1]
            counter_dict[nums[i - k +1]] -= 1
            if counter_dict[nums[i - k +1]] == 0:
                del counter_dict[nums[i - k +1]]
        return max(sum_list) if sum_list else 0
    
#! I actually made it right, cannot believe it
# @lc code=end

File: 2841.maximum-sum-of-almost-unique-subarray.java
/*
 * @lc app=leetcode id=2841 lang=java
 *
 * [2841] Maximum Sum of Almost Unique Subarray
 */

import java.util.HashMap;
import java.util.List;
import java.util.Map;

// @lc code=start
class Solution {
    public long maxSum(List<Integer> nums, int m, int k) {
        long result = 0;
        long sum = 0;
        int n = nums.size();
        Map<Integer, Integer> map = new HashMap<>();

        for(int i = 0; i < n; i++){
            sum += nums.get(i);
            map.put(nums.get(i), map.getOrDefault(nums.get(i), 0) + 1);

            if(i < k -1){
                continue;
            }

            if(map.size() >= m){
                result = Math.max(result, sum);
            }

            sum -= nums.get(i - k + 1);
            int count = map.get(nums.get(i - k + 1));
            if(count == 1){
                map.remove(nums.get(i - k + 1));
            } else {
                map.put(nums.get(i - k + 1), count - 1);
            }

        }
        return result;
    }
}
// @lc code=end

